compiler/modules/types/object.bal

```diff
@@ -0,0 +1,64 @@
+public type Member record {|
+    string name;
+    SemType valueTy;
+    "field"|"method" kind;
+|};
+
+SemType MEMBER_KIND_FIELD = stringConst("field");
+SemType MEMBER_KIND_METHOD = stringConst("method");
+
+public class ObjectDefinition {
+    *Definition;
+    private MappingDefinition mappingDefn = new();
+
+    public function getSemType(Env env) returns SemType {
+        return self.createSemType(self.mappingDefn.getSemType(env));
+    }
+
+    public function define(Env env, Member[] members) returns SemType {
+        CellField[] fieldCells = from Member member in members where member.kind == "field" select fieldMember(env, member);
+        CellField[] methodCells = from Member member in members where member.kind == "method" select methodMember(env, member);
+        SemType mappingType = self.mappingDefn.define(env, [...fieldCells, ...methodCells],
+                                                      restMemberType(env));
+        return self.createSemType(mappingType);
+    }
+
+    private function createSemType(SemType mappingType) returns SemType {
+        Bdd bdd = <Bdd>subtypeData(mappingType, BT_MAPPING);
+        return createBasicSemType(BT_OBJECT, bdd);
+    }
+}
+
+function restMemberType(Env env) returns CellSemType {
+    SemType fieldMemberType = defineMappingTypeWrapped(new, env, [{ name: "value", ty: ANY },
+                                                                  { name: "kind", ty: MEMBER_KIND_FIELD }],
+                                                       NEVER);
+    SemType methodMemberType = defineMappingTypeWrapped(new, env, [{ name: "value", ty: FUNCTION, ro: true },
+                                                                   { name: "kind", ty: MEMBER_KIND_METHOD }],
+                                                        NEVER);
+    return cellContaining(env, union(fieldMemberType, methodMemberType));
```

Comment:
@jclark It turned out that every parameter in the function type is also represented as individual cells ([ref](https://github.com/ballerina-platform/nballerina/blob/0331ffaf618ed6e5ece4efa405e8712105cdeabc/compiler/modules/front/resolveTypes.bal#L336C40-L336C40)) and list type also expect each member to be a cell. Should I still change the mapping type to such that members are not cells?

Reply:
Draft PR with changes that will be needed #1219